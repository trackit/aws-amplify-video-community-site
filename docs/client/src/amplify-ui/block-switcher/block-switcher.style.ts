import {css} from "emotion";

export const hostStyle = css`
  display: block;
  margin-bottom: 1.5rem;
  background-color: white;
  border: 0.0625rem solid var(--border-color);
  border-radius: 0.25rem;
  > pre {
    display: none;
  }

  amplify-code-block {
    margin-top: 0;
  }
`;

export const activeTabStyle = css`
  font-weight: 700;
  border-bottom: 0.25rem solid var(--amplify-primary-color);
`;

export const tabStyle = css`
  border-top-right-radius: 0.25rem;
  border-top-left-radius: 0.25rem;
  height: 100%;
  appearance: none;
  background-color: transparent;
  padding: 0.75rem 1rem 0.5rem;
  cursor: pointer;
  font-size: 0.875rem;
`;

export const tabContainerStyle = css`
  border-top-right-radius: 0.25rem;
  border-top-left-radius: 0.25rem;
  border-bottom: 0.0625rem solid var(--border-color);
`;

export const contentStyle = css`
  padding: 1rem;
  border-bottom-right-radius: 0.25rem;
  border-bottom-left-radius: 0.25rem;

  > amplify-block {
    display: none;
    > pre:last-child {
      margin-bottom: 0;
    }
  }
`;
