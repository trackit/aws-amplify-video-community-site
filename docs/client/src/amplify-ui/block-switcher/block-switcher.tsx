import {Component, Host, h, Prop, State, Element, Watch} from "@stencil/core";
import {
  hostStyle,
  activeTabStyle,
  tabStyle,
  tabContainerStyle,
  contentStyle,
} from "./block-switcher.style";
import {css} from "emotion";
import {pageContext} from "../../docs-ui/page/page.context";
import {
  SelectedTabHeadings,
  SetNewSelectedTabHeadings,
} from "../../docs-ui/page/page.types";

const BLOCK_TAG_NAME = "amplify-block";

@Component({tag: "amplify-block-switcher", shadow: false})
export class AmplifyBlockSwitcher {
  @Element() el: HTMLElement;

  /** list of previously tab headings in order of priority, passed from global provider */
  @Prop() readonly selectedTabHeadings: SelectedTabHeadings;
  /** tack on a new tab heading at highest priority */
  @Prop() readonly setNewSelectedTabHeadings: SetNewSelectedTabHeadings;
  /** increments whenever the platform changes and we need to refresh the tabHeadings */
  @Prop() readonly alwaysRerenderBlockSwitcher: number;

  @State() activeChildI = 0;

  tabHeadings: string[] = [];

  componentWillRender() {
    this.gatherHeadings();
  }

  // recursively traverse the DOM tree to gather the language names from all child code blocks.
  // when originally constructed, the code blocks are briefly direct children of the block-switcher, living in the shadow DOM,
  // but then they are moved into slots, which are further down in the component, and thus we must recurse down to reach them.
  recursivelyFindBlocks(el: HTMLElement) {
    if (el.matches(BLOCK_TAG_NAME)) {
      // somehow this doesn't cause a rerender each time a heading is pushed, just at the end
      this.tabHeadings.push(((el as any) as {name: string}).name);
      return;
    }
    const children = Array.from(el.children);
    for (const child of children) {
      this.recursivelyFindBlocks(child as HTMLElement);
    }
  }

  gatherHeadings() {
    // gather tab headings from child `amplify-block` attrs
    this.tabHeadings = [];
    this.recursivelyFindBlocks(this.el);

    // default to the first tab
    this.activeChildI = 0;

    if (this.tabHeadings?.length > 0 && this.selectedTabHeadings?.length > 0) {
      // iterate through previous selections, and see if we have any shared
      for (let i = 0; i < this.selectedTabHeadings.length; i++) {
        const current = this.selectedTabHeadings[i];
        if (this.tabHeadings.includes(current)) {
          // if so, let's set its index in state!
          this.activeChildI = this.tabHeadings.indexOf(current);
          break;
        }
      }
    }
  }

  @Watch("selectedTabHeadings")
  synchronizeSameTabs(): void {
    for (const e of this.selectedTabHeadings) {
      const eIndexInTabHeadings = this.tabHeadings.indexOf(e);
      if (eIndexInTabHeadings >= 0) {
        this.activeChildI = eIndexInTabHeadings;
        break;
      }
    }
  }

  createActiveChildISetter = (i: number) => (_event: Event): void => {
    this.activeChildI = i;
    const current = this.tabHeadings[i];
    this.setNewSelectedTabHeadings(current);
  };

  render() {
    // + 1 because CSS child indices start at 1
    const revealStyle = css`
      amplify-block:nth-of-type(${this.activeChildI + 1}) {
        display: initial;
      }
    `;

    return (
      <Host
        class={{
          [hostStyle]: true,
          [revealStyle]: true,
        }}
      >
        <div class={tabContainerStyle}>
          {this.tabHeadings?.map((e, i) => {
            return (
              <button
                onClick={this.createActiveChildISetter(i)}
                class={{
                  [activeTabStyle]: this.activeChildI === i,
                  [tabStyle]: true,
                }}
                key={e}
              >
                {e}
              </button>
            );
          })}
        </div>
        <div class={contentStyle}>
          <slot />
        </div>
      </Host>
    );
  }
}

pageContext.injectProps(AmplifyBlockSwitcher, [
  "alwaysRerenderBlockSwitcher",
  "selectedTabHeadings",
  "setNewSelectedTabHeadings",
]);
