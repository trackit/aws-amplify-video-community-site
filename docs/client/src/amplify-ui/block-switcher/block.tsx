import {Component, Host, h, Prop, Element} from "@stencil/core";

@Component({tag: "amplify-block", shadow: false})
export class AmplifyCodeBlock {
  /** Block name */
  @Prop() readonly name?: string;

  render() {
    return (
      <Host name={this.name}>
        <slot></slot>
      </Host>
    );
  }
}
