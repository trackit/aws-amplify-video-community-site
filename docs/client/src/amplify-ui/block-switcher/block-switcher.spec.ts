import {AmplifyBlockSwitcher} from "./block-switcher";
import {AmplifyCodeBlock} from "./block";
import {newSpecPage} from "@stencil/core/testing";

describe("amplify-block-switcher", () => {
  describe("Component logic", () => {
    it("should recursively gather all block names as headings", async () => {
      const page = await newSpecPage({
        components: [AmplifyBlockSwitcher, AmplifyCodeBlock],
        html: `
          <amplify-block-switcher>
            <div id="red herring">
              <amplify-blurk name="nope"></amplify-blurk>
            </div>
            <div>
              <amplify-block name="test 1"></amplify-block>
              <amplify-block name="test 2"></amplify-block>
            </div>
          </amplify-block-switcher>`,
      });
      const blockSwitcher = page.rootInstance;
      expect(blockSwitcher.tabHeadings).toEqual(["test 1", "test 2"]);
    });
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyBlockSwitcher],
            html: `<amplify-block-switcher />`,
          })
        ).root,
      ).toMatchSnapshot();
    });

    it("should render with contained blocks", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyBlockSwitcher],
            html: `<amplify-block-switcher><amplify-block></amplify-block></amplify-block-switcher>`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
