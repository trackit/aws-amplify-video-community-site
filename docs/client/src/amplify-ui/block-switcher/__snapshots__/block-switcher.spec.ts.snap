// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-block-switcher Render logic should render 1`] = `
<amplify-block-switcher class="css-1dfg3q css-6d0h3e">
  <!---->
  <div class="css-13kxqe8"></div>
  <div class="css-1fwr3l6"></div>
</amplify-block-switcher>
`;

exports[`amplify-block-switcher Render logic should render with contained blocks 1`] = `
<amplify-block-switcher class="css-1dfg3q css-6d0h3e">
  <!---->
  <div class="css-13kxqe8">
    <button class="css-18ssisc css-si1pwq"></button>
  </div>
  <div class="css-1fwr3l6">
    <amplify-block></amplify-block>
  </div>
</amplify-block-switcher>
`;
